### PR TITLE
Allow fonts other than the system font

### PR DESCRIPTION
**Pull request by piepie62**
Slight change to allow code reuse of many font functions.

### CODE REVIEW

File: libctru/source/font.c [hunk 1]
```diff
@@ -26,17 +25,34 @@ Result fontEnsureMapped(void)
...
+	font->finf.cwdh = (CWDH_s*)((u32)(font->finf.cwdh) + (u32) font);
+	CWDH_s* cwdh;
+	for (cwdh = font->finf.cwdh; cwdh->next; cwdh = cwdh->next)
+		cwdh->next = (CWDH_s*)((u32)(cwdh->next) + (u32) font);
+}
+
+int fontGlyphIndexFromCodePoint(CFNT_s* font, u32 codePoint)
+{
+	if (!font)
+		font = fontGetSystemFont();
```
[the remark this answers is not in the document]
Comment (piepie62) [reply libctru/source/font.c]: While that's true, I thought this was cleaner than declaring the `extern` variable. If you would like this better, I will change it.

File: libctru/include/3ds/font.h [hunk 2]
```diff
@@ -157,49 +157,79 @@ enum
 /// Ensures the shared system font is mapped.
 Result fontEnsureMapped(void);
 
-/// Retrieves the font information structure of the shared system font.
-static inline FINF_s* fontGetInfo(void)
+/**
+ * @brief Fixes the pointers internal to a just-loaded font
+ * @param font Font to fix
+ * @remark Should never be run on the system font, and only once
+ * 		   on any other font.
```
Comment (fincs) [inline libctru/include/3ds/font.h]: Spaces for alignment

File: libctru/include/3ds/font.h [hunk 3]
```diff
@@ -157,49 +157,79 @@ enum
 /// Ensures the shared system font is mapped.
 Result fontEnsureMapped(void);
 
-/// Retrieves the font information structure of the shared system font.
-static inline FINF_s* fontGetInfo(void)
+/**
+ * @brief Fixes the pointers internal to a just-loaded font
+ * @param font Font to fix
+ * @remark Should never be run on the system font, and only once
+ * 	       on any other font.
```
Comment (fincs) [inline libctru/include/3ds/font.h]: Move to same line as above.

File: libctru/source/font.c [hunk 4]
```diff
@@ -26,17 +25,34 @@ Result fontEnsureMapped(void)
...
-int fontGlyphIndexFromCodePoint(u32 codePoint)
+void fontFixPointers(CFNT_s* font)
 {
-	int ret = g_sharedFont->finf.alterCharIndex;
+	font->finf.tglp = (TGLP_s*)((u32)(font->finf.tglp) + (u32) font);
+	font->finf.tglp->sheetData = (u8*)((u32)(font->finf.tglp->sheetData) + (u32) font);
+
+	font->finf.cmap = (CMAP_s*)((u32)(font->finf.cmap) + (u32) font);
+	CMAP_s* cmap = font->finf.cmap;
+	for (; cmap->next; cmap = cmap->next)
```
Comment (fincs) [inline libctru/source/font.c]: `for (CMAP_s* cmap = font->finf.cmap; cmap->next; cmap = cmap->next)`

File: libctru/source/font.c [hunk 5]
```diff
@@ -26,17 +25,34 @@ Result fontEnsureMapped(void)
...
+		cmap->next = (CMAP_s*)((u32)(cmap->next) + (u32) font);
+
+	font->finf.cwdh = (CWDH_s*)((u32)(font->finf.cwdh) + (u32) font);
+	CWDH_s* cwdh;
+	for (cwdh = font->finf.cwdh; cwdh->next; cwdh = cwdh->next)
```
Comment (fincs) [inline libctru/source/font.c]: `for (CWDH_s* cwdh = font->finf.cwdh; cwdh->next; cwdh = cwdh->next)`

File: libctru/source/font.c [hunk 6]
```diff
@@ -26,17 +25,34 @@ Result fontEnsureMapped(void)
...
+		cwdh->next = (CWDH_s*)((u32)(cwdh->next) + (u32) font);
+}
+
+int fontGlyphIndexFromCodePoint(CFNT_s* font, u32 codePoint)
+{
+	if (!font)
+		font = g_sharedFont;
+	int ret = font->finf.alterCharIndex;
```
Comment (fincs) [inline libctru/source/font.c]: Add another `font` null check here (can happen if g_sharedFont is null) and return -1 in that case.

File: libctru/source/font.c [hunk 7]
```diff
@@ -67,26 +83,31 @@ int fontGlyphIndexFromCodePoint(u32 codePoint)
 	return ret;
 }
 
-charWidthInfo_s* fontGetCharWidthInfo(int glyphIndex)
+charWidthInfo_s* fontGetCharWidthInfo(CFNT_s* font, int glyphIndex)
 {
+	if (!font)
+		font = g_sharedFont;
 	charWidthInfo_s* info = NULL;
```
Comment (fincs) [inline libctru/source/font.c]: Same here; return NULL.

File: libctru/source/font.c [hunk 8]
```diff
@@ -67,26 +83,31 @@ int fontGlyphIndexFromCodePoint(u32 codePoint)
...
 	return info;
 }
 
-void fontCalcGlyphPos(fontGlyphPos_s* out, int glyphIndex, u32 flags, float scaleX, float scaleY)
+void fontCalcGlyphPos(fontGlyphPos_s* out, CFNT_s* font, int glyphIndex, u32 flags, float scaleX, float scaleY)
 {
-	FINF_s* finf = &g_sharedFont->finf;
+	if (!font)
+		font = g_sharedFont;
+	FINF_s* finf = &font->finf;
```
Comment (fincs) [inline libctru/source/font.c]: Same here, and return.

**Pull request merged**